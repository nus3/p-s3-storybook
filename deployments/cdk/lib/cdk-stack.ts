import * as cdk from '@aws-cdk/core'
import * as s3 from '@aws-cdk/aws-s3'
import * as iam from '@aws-cdk/aws-iam'
import { Duration } from '@aws-cdk/core'

interface s3Config {
  bucketName: string
  name: string
}

interface StageContext {
  iam: string
  description: string
  s3: s3Config[]
}

export class CdkStack extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props)

    const env: string = this.node.tryGetContext('env') || 'Dev'
    const revision: string = this.node.tryGetContext('revision') || ''

    const context: StageContext =
      env === 'Test'
        ? JSON.parse(this.node.tryGetContext(env))
        : this.node.tryGetContext(env)

    this.validateEnvironment(env, revision, context)

    this.createS3Buckets(context)
  }

  private validateEnvironment(
    env: string,
    revision: string,
    context: StageContext
  ): void {
    if (!context.s3) {
      throw new Error(
        `error: invalid s3 context ${JSON.stringify({
          env,
          revision,
          context,
        })}`
      )
    }

    context.s3.forEach((config) => {
      if (!(config.name && config.bucketName)) {
        throw new Error(
          `error: invalid s3 context ${JSON.stringify({
            env,
            revision,
            context,
          })}`
        )
      }
    })
  }

  private createS3Buckets(context: StageContext): s3.Bucket[] {
    const buckets: s3.Bucket[] = []

    context.s3.forEach((config) => {
      const bucket: s3.Bucket = new s3.Bucket(this, config.name, {
        versioned: false,
        bucketName: config.bucketName,
        publicReadAccess: true,
        removalPolicy: cdk.RemovalPolicy.DESTROY,
        lifecycleRules: [
          {
            expiration: Duration.days(14),
            prefix: 'branches/',
          },
        ],
        websiteIndexDocument: 'index.html',
      })

      const policyStatement: iam.PolicyStatement = new iam.PolicyStatement({
        effect: iam.Effect.ALLOW,
        actions: ['s3:ListBucket', 's3:PutObject', 's3:GetBucketAcl'],
        resources: [
          `arn:aws:s3:::${config.bucketName}`,
          `arn:aws:s3:::${config.bucketName}/*`,
        ],
        principals: [new iam.AccountPrincipal(context.iam)],
      })

      bucket.addToResourcePolicy(policyStatement)

      buckets.push(bucket)
    })

    return buckets
  }
}
