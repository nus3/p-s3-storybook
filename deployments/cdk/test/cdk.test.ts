import { SynthUtils } from '@aws-cdk/assert'

import * as cdk from '@aws-cdk/core'
import { CdkStack } from '../lib/cdk-stack'

test('Snapshot Test', () => {
  const account: string = '000000000000'
  const region: string = 'ap-northeast-1'
  const context: any = {
    description: 'description',
    iam: 'account-id',
    s3: [
      {
        bucketName: 'storybook.test.p-s3-storybook',
        name: 'p-s3-storybook_Bucket',
      },
    ],
  }
  const app: cdk.App = new cdk.App({
    context: { env: 'Test', Test: JSON.stringify(context) },
  })
  const stack = new CdkStack(app, 'TestStack', {
    env: { account, region },
  })
  const cfn: any = SynthUtils.toCloudFormation(stack)
  expect(cfn).toMatchSnapshot()
})
