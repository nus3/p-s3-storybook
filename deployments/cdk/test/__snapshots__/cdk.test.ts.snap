// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
Object {
  "Resources": Object {
    "ps3storybookBucketBAA5F972": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "BucketName": "storybook.test.p-s3-storybook",
        "LifecycleConfiguration": Object {
          "Rules": Array [
            Object {
              "ExpirationInDays": 14,
              "Prefix": "branches/",
              "Status": "Enabled",
            },
          ],
        },
        "WebsiteConfiguration": Object {
          "IndexDocument": "index.html",
        },
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "ps3storybookBucketPolicy514DFB0B": Object {
      "Properties": Object {
        "Bucket": Object {
          "Ref": "ps3storybookBucketBAA5F972",
        },
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "s3:GetObject",
              "Effect": "Allow",
              "Principal": "*",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    Object {
                      "Fn::GetAtt": Array [
                        "ps3storybookBucketBAA5F972",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
            Object {
              "Action": Array [
                "s3:ListBucket",
                "s3:PutObject",
                "s3:GetBucketAcl",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::account-id:root",
                    ],
                  ],
                },
              },
              "Resource": Array [
                "arn:aws:s3:::storybook.test.p-s3-storybook",
                "arn:aws:s3:::storybook.test.p-s3-storybook/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
  },
}
`;
